---
title: "Filtering Step 1 - SNP and Individual Quality"
author: "Anita Wray"
date: "`r Sys.Date()`"
output: html_document
---
### *This script plots PCAs of data generated by STACKS & filtered in R.*

```{r setup, message=FALSE, warning=FALSE, include=FALSE}
knitr::opts_knit$set(root.dir = 'C:/Users/Anita\ Wray/Desktop/AwrayRockfish/populations/refmap/black')
```


```{r packages, include=FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
knitr::opts_chunk$set(engine.opts = list(bash = "-l"))

source("C:/Users/Anita Wray/Desktop/AwrayRockfish/scripts/libraries.R")
source("C:/Users/Anita Wray/Desktop/AwrayRockfish/scripts/ggplot.R")
source("C:/Users/Anita Wray/Desktop/AwrayRockfish/scripts/VCFfilterstats.R")
source("C:/Users/Anita Wray/Desktop/AwrayRockfish/scripts/xtrafunctions.R")

pkgs <- c('ggplot2', 'plotly', 'adegenet', 'graphics', 'vcfR', 'tidyr', 'dplyr', 'hierfstat', 'poppr', 'graphics', 'viridis', 'pegas', 'factoextra')
lapply(pkgs, library, character.only = TRUE)
```

---
## Filtering scheme 3

### Step 1: Filter loci

Filter loci with minimum genotype depth \< 5, then remove loci with mean minimum depth \< 15, filter loci with minor allele frequency \< 5%, finally apply genotype call rate \> 80%.

```{bash filter loci, include=FALSE}

vcftools --vcf populations.snps.vcf --out black_minDP5meanDP15miss80maf05 --minDP 5 --min-meanDP 15 --max-missing 0.80 --maf 0.05 --recode --recode-INFO-all

# stats combinedspecies
vcftools --vcf black_minDP5meanDP15miss80maf05.recode.vcf --out black_minDP5meanDP15miss80maf05 --depth
vcftools --vcf black_minDP5meanDP15miss80maf05.recode.vcf --out black_minDP5meanDP15miss80maf05 --site-mean-depth
vcftools --vcf black_minDP5meanDP15miss80maf05.recode.vcf --out black_minDP5meanDP15miss80maf05 --site-quality
vcftools --vcf black_minDP5meanDP15miss80maf05.recode.vcf --out black_minDP5meanDP15miss80maf05 --missing-indv
vcftools --vcf black_minDP5meanDP15miss80maf05.recode.vcf --out black_minDP5meanDP15miss80maf05 --missing-site
vcftools --vcf black_minDP5meanDP15miss80maf05.recode.vcf --out black_minDP5meanDP15miss80maf05 --het

```

Decompose variants and retain only SNPs.

```{bash retain SNPs, include=FALSE}

vcflib vcfallelicprimitives black_minDP5meanDP15miss80maf05.recode.vcf --keep-info --keep-geno > SNPs.vcf

vcftools --vcf SNPs.vcf --out black_minDP5meanDP15miss80maf05.SNPs --remove-indels --recode --recode-INFO-all

vcftools --vcf black_minDP5meanDP15miss80maf05.SNPs.recode.vcf --out black_minDP5meanDP15miss80maf05.SNPs  --missing-indv

```

### Step 2: Identify individuals with \> 25% missing data

```{r 25 miss table, echo=FALSE}

imissing <- read.table("black_minDP5meanDP15miss80maf05.imiss",
                       header = TRUE, stringsAsFactors = FALSE)

ggplot(imissing, aes(x = F_MISS)) +
  geom_histogram(binwidth = .005, color = "black", fill = "grey95") +
  geom_vline(aes(xintercept = mean(F_MISS, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 0.25),
                 color = "darkblue", linetype = "dashed", size = 1) +
  scale_x_continuous(limits = c(0, 1)) +
  labs(x = "missing data per indv") +
  theme_standard

LQ_indv <- imissing %>%
  filter(F_MISS > 0.25) %>%
  select(INDV)

write.table(LQ_indv, "LQ_Ind_FIL-3", 
            col.names = FALSE, row.names = FALSE, quote = FALSE)
LQ_indv
```

Remove flagged individuals

```{bash remove missing indv, include=FALSE}

vcftools --vcf black_minDP5meanDP15miss80maf05.recode.vcf --out black_FIL-3 --remove LQ_Ind_FIL-3 --recode --recode-INFO-all

# stats black
vcftools --vcf black_FIL-3.recode.vcf --out black_FIL-3 --depth
vcftools --vcf black_FIL-3.recode.vcf --out black_FIL-3 --site-mean-depth
vcftools --vcf black_FIL-3.recode.vcf --out black_FIL-3 --site-quality
vcftools --vcf black_FIL-3.recode.vcf --out black_FIL-3 --missing-indv
vcftools --vcf black_FIL-3.recode.vcf --out black_FIL-3 --missing-site
vcftools --vcf black_FIL-3.recode.vcf --out black_FIL-3 --het
```

### Plot stats for filtered data set

```{r stats FIL 3, echo=FALSE, fig.height=20, fig.width=10, message=FALSE, warning=FALSE}

# load stats files ----
ind_stats_FIL_3 <- read.ind.stats(dir = 'C:/Users/Anita\ Wray/Desktop/AwrayRockfish/populations/refmap/black', 
                                  vcf = "black_FIL-3")

loc_stats_FIL_3 <- read.loc.stats(dir = 'C:/Users/Anita\ Wray/Desktop/AwrayRockfish/populations/refmap/black',
                                  vcf = "black_FIL-3")

# plot missing data per indv ----
p1 <- ggplot(ind_stats_FIL_3, aes(x = `MISS_black_FIL-3`)) +
  geom_histogram(binwidth = .01, color = "black", fill = "grey95") +
  geom_vline(aes(xintercept = mean(`MISS_black_FIL-3`, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 0.5),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "missing data per indv") +
  theme_standard

# plot Fis per indv ----
p2 <- ggplot(ind_stats_FIL_3, aes(x = `Fis_black_FIL-3`)) +
  geom_histogram(binwidth = .01, color = "black", fill = "grey95") +
  geom_vline(aes(xintercept = mean(`Fis_black_FIL-3`, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 0),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "Fis per indv") +
  theme_standard

# plot read depth per indv ----
p3 <- ggplot(ind_stats_FIL_3, aes(x = `MEAN_DEPTH_black_FIL-3`)) +
  geom_histogram(binwidth = 10, color = "black", fill = "grey95") +
  geom_vline(aes(xintercept = mean(`MEAN_DEPTH_black_FIL-3`, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 20),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "mean read depth per indv") +
  theme_standard

# plot depth vs missing ----
p4 <- ggplot(ind_stats_FIL_3, aes(x = `MEAN_DEPTH_black_FIL-3`, y = `MISS_black_FIL-3`)) +
  geom_point() +
  geom_vline(aes(xintercept = mean(`MEAN_DEPTH_black_FIL-3`, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 20),
                 color = "darkblue", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = mean(`MISS_black_FIL-3`, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = 0.5),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "mean depth per indv", y = "% missing data") +
  theme_standard

# plot Fis vs missing data per indv ----
p5 <- ggplot(ind_stats_FIL_3, aes(x = `Fis_black_FIL-3`, y = `MISS_black_FIL-3`)) +
  geom_point() +
  geom_vline(aes(xintercept = mean(`Fis_black_FIL-3`, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 0),
                 color = "darkblue", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = mean(`MISS_black_FIL-3`, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = 0.5),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "Fis per indv", y = "% missing data") +
  theme_standard

# plot Fis vs mean depth per indv ----
p6 <- ggplot(ind_stats_FIL_3, aes(x = `Fis_black_FIL-3`, y = `MEAN_DEPTH_black_FIL-3`)) +
  geom_point() +
  geom_vline(aes(xintercept = mean(`Fis_black_FIL-3`, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 0),
                 color = "darkblue", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = mean(`MEAN_DEPTH_black_FIL-3`, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = 20),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "Fis per indv", y = "mean depth per indv") +
  theme_standard

# plot distribution missing data per locus ----
p7 <- ggplot(loc_stats_FIL_3, aes(x = `MISS_black_FIL-3`)) +
  geom_histogram(binwidth = 0.01, color = "black", fill = "grey95") +
  geom_vline(aes(xintercept = mean(`MISS_black_FIL-3`, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 0.1),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "% missing data per locus") +
  theme_standard

# plot distribution mean read depth ----
p8 <- ggplot(loc_stats_FIL_3, aes(x = `MEAN_DEPTH_black_FIL-3`)) +
  geom_histogram(binwidth = 5, color = "black", fill = "grey95") +
  geom_vline(aes(xintercept = mean(`MEAN_DEPTH_black_FIL-3`, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 20),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "mean read depth per locus") +
  theme_standard

# plot read depth vs missing data ----
p9 <- ggplot(loc_stats_FIL_3, aes(x = `MEAN_DEPTH_black_FIL-3`, y = `MISS_black_FIL-3`)) +
  geom_point() +
  geom_vline(aes(xintercept = mean(`MEAN_DEPTH_black_FIL-3`, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 20),
                 color = "darkblue", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = mean(`MISS_black_FIL-3`, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = 0.1),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "mean depth per locus", y = "% missing data") +
  theme_standard

# plot no of SNPs per locus ----
p10 <- loc_stats_FIL_3 %>%
  dplyr::count(CHR) %>%
  ggplot(aes(x = n)) +
  geom_histogram(binwidth = 1, color = "black", fill = "grey95") + 
  labs(x = "number of SNPs per locus") +
  theme_standard

temp <- loc_stats_FIL_3 %>%
  dplyr::count(CHR)

# plot number of SNPs per contig vs. mean depth ----
p11 <- left_join(temp, loc_stats_FIL_3) %>%
  ggplot() +
  geom_point(aes(x = n, y = `MEAN_DEPTH_black_FIL-3`)) +
  labs(x = "number of SNPs per contig", y = "mean depth") +
  theme_standard

# plot depth vs SNP quality ----
site_qual <- read.table("black_FIL-3.lqual", 
                        header = TRUE, stringsAsFactors = FALSE) %>%
  mutate(PROB = 10^(-QUAL/10))

temp <- data.frame(loc_stats_FIL_3$`MEAN_DEPTH_black_FIL-3`, site_qual$QUAL) %>%
  dplyr::rename(depth = loc_stats_FIL_3..MEAN_DEPTH_black_FIL.3., qual = site_qual.QUAL)

p12 <- ggplot(temp, aes(x = depth, y = qual)) +
  geom_point(size = 1) +
  geom_vline(aes(xintercept = mean(depth, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_vline(aes(xintercept = 20),
                 color = "darkblue", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = mean(qual, na.rm = TRUE)),
                 color = "red", linetype = "dashed", size = 1) +
  geom_hline(aes(yintercept = 20),
                 color = "darkblue", linetype = "dashed", size = 1) +
  labs(x = "mean depth per locus", y = "SNP quality") +
  theme_standard

m4 <- multiplot(p1, p2, p3, p4, p5, p6, p7, p8, p9, p10, p11, p12, cols=2)

```

### **Data set contains `r nrow(ind_stats_FIL_3)` individuals and `r nrow(loc_stats_FIL_3)` loci.**

```{r plot f3, include=FALSE}

# File to be used
my_vcf <-  read.vcfR("black_FIL-3.recode.vcf")
my_genind <- vcfR2genind(my_vcf)

#Complementary Data
extra_data <- read.delim("C:/Users/Anita\ Wray/Desktop/AwrayRockfish/Rpopassignments/black_popmap.txt")

name_vec <- indNames(my_genind)
pop_vec <- extra_data[extra_data$Sample %in% ind_stats_FIL_3$INDV,]$Population
pop(my_genind) <- pop_vec
Adata_scaled <- scaleGen(my_genind, NA.method = "mean")
pca_A <- dudi.pca(Adata_scaled,cent = TRUE,scale = FALSE,scannf = FALSE, nf = 20)
PCA_df <- pca_A$li
PCA_df$SampleID <- name_vec
pca_info <- get_eigenvalue(pca_A)
PCA_df$Population <- pop_vec

pca_black3 <- ggplot(data = PCA_df, aes(x = Axis1, y = Axis2, color = Population)) +
  geom_point(size = 2.5, alpha = 0.6) +
  xlab(paste("Axis 1 (", round(pca_info$variance.percent[[1]], 3), "%)", sep = "")) +
  ylab(paste("Axis 2 (", round(pca_info$variance.percent[[2]], 3), "%)", sep = "")) +
  theme_Publication() + ggtitle('black Fil3') +
  geom_text(aes(label=SampleID), hjust='inward', vjust='inward')
  
ggplotly(pca_black3)
```

`r ggplotly(pca_black3)`