---
title: "Filtering Step 6 - LD"
author: "Anita Wray"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_knit$set(root.dir = 'C:/Users/Anita\ Wray/Desktop/AwrayRockfish/populations/refmap/black')

# Load libraries
library(vcfR)
library(genetics)
library(tidyr)
library(dplyr)
library(ggplot2)
library(janitor)
require(snpStats)
library(LDheatmap)
library(viridis)
library(scales)
library(RColorBrewer)
library(scatterpie)
library(stringr)
```

# The purpose of this script is to use the genetics package to calculate LD on each chromosome separately. Then, take a directory of vcf files (split by chromosome), calculate LD within each chromosome, and save the output as a heatmap. Finally, plot LD decay plot and summary statistics per chromosome.
# To run the code, put all of your vcf files into a single directory.
# Script by Eleni Petrou, 6/19/2020, using code modified from": https://sfustatgen.github.io/LDheatmap/articles/vcfOnLDheatmap.html

```{r Calculate LD}

# Specify the names of data files used - they all end in .vcf
fileNames <- Sys.glob("*.vcf") #this is R's version of a wildcard

################################################################################
# read in vcf files in directory using vcfR, and start data processing
i = 1

for (fileName in fileNames) {
  print(i)
  
  
  vcf_data <- read.vcfR(fileName, verbose = F)
  
  vcf_data #take a peek
  #head(vcf_data)
  
  #save metadata as dataframe - you will need this later for plotting
  vcf_df <- as.data.frame(vcf_data@fix)
  #head(vcf_df) #check
  #class(vcf_df) #should be dataframe
  
  # create a new variable, SNPID, to label the snps
  vcf_df <- vcf_df %>%
    mutate(SNPID = paste0(CHROM,"_",POS))
  
  head(vcf_df)
  
  #use vcfR to extract the genotypes from the vcf file --> make matrix
  gt <- extract.gt(vcf_data, return.alleles = TRUE)
  #gt[1:10, 1:10] #take a peek
  class(gt) #should be matrix
  
  # Set the row names to the alphanumeric SNPID
  rownames(gt) <- vcf_df$SNPID
  
  #############
  #Prepare data for genetics package and LD calculation
  #transpose the genotype matrix and make it into a data frame
  gt_df <- data.frame(t(gt))
  
  #gt_df[1:10, 1:10] #take a peek
  class(gt_df) #should be dataframe
  
  # change "." to NAs
  gt_df[gt_df == "."] <- NA
  gt_df[1:10, 1:10] #take a peek
  class(gt_df) #should be dataframe
  
  # Now that you have a dataframe of genotypes, turn it into a genetics object
  gt_genetics <- makeGenotypes(gt_df)
  
  # Run the LD test and hope for the best
  output<- LD(gt_genetics)
  #class(output) # the output is a series of horrible, nested triangular matrices. FUUUCK
  #head(output)
  
  test <- as.data.frame(as.table((output$`R^2`))) #saved the output as a dataframe
  
  
  ### get a dataframe into tidy format for plotting 
  #head(test)
  
  ## add metadata to the dataframe, regarding the position of the SNP in the chromosome
  
  plot_df <- test %>%
    drop_na() %>% #drop all the dumb NA's from the rectangular matrix
    left_join(vcf_df, by = c("Var1" = "SNPID")) %>%
    dplyr::select(Var1, Var2, Freq, CHROM, POS) %>% #add the position of the first locus
    rename(r2 = Freq, locus1_pos = POS) #rename the columns
  
  plot_df2 <- plot_df %>%
    left_join(vcf_df, by = c("Var2" = "SNPID"))%>% #add the position of the second locus
    dplyr::select(Var1, Var2, r2, CHROM.x, locus1_pos, POS) %>%
    rename(CHROM = CHROM.x, locus2_pos = POS) #rename the columns
  
  #head(plot_df2) #take a peek
  
  # save the positions as numeric objects (and not factors)
  plot_df2$locus1_pos <- as.numeric(as.character(plot_df2$locus1_pos)) 
  plot_df2$locus2_pos <- as.numeric(as.character(plot_df2$locus2_pos))
  
  #calculate the absolute distance in bp between SNPS on a chromosome
  plot_df3 <- plot_df2 %>%
    mutate(distance = abs(locus1_pos - locus2_pos))
  
  #head(plot_df3) #take a peek
  #################################
  #Plot the output in ggplot
  pdf(paste(fileName, "LD_decay.pdf")) 
  
  # LD decay
  myplot <- ggplot(plot_df3)+
    geom_point(aes(x = distance, y =r2), color = "skyblue2")+
    #geom_smooth(aes(x = distance, y =r2), method = "gam")+
    xlab("Distance(base pairs)") +
    ylab(bquote("Linkage disequilibrium"~(italic(R^2))))+
    theme_bw()
  myplot
  print(myplot)
  
  dev.off()
  
  i = i+1
  
  #################################
  #Save the output to a text file
  
  write.table(plot_df3, file = paste(fileName, "LD_results.txt"), append = FALSE, quote = FALSE, sep = "\t",
              eol = "\n", na = "NA", dec = ".", row.names = FALSE,
              col.names = TRUE)
  
  
}
```

```{r Plot Heatmap per Chromosome}

# Specify the names of data files used
fileNames <- Sys.glob("*.vcf") #this is R's version of a wildcard


################################################################################
#read in vcf files in directory using vcfR, and start data processing

for (fileName in fileNames) {
  
  snp <- read.vcfR(fileName)
  snp@gt[1:5, 1:5]
  
  #save vcf metadata as dataframe - you will need this information later for plotting
  vcf_df <- as.data.frame(snp@fix)
  head(vcf_df) #check
  class(vcf_df) #should be dataframe
  
  ################################################################################
  # Process the genotypes to get them into SNPMatrix format
  # select all genotype column values from vcf file and first FORMAT column
  gt_temp <- snp@gt
  gt_temp[1:10,1:10] #take a peek
  
  snpMat <- t(gt_temp) #transpose this matrix
  snpMat[1:5, 1:5]
  
  #Convert the matrix of genotypes to a numeric matrix in which genotypes are 
  #coded as 0, 1 or 2 copies of the minor allele.
  
  #define a function to convert the value of genotypes into 0,1,2
  convertToNumeric <- function(x){
    gdat <- matrix(NA,nrow = nrow(x), ncol = ncol(x))
    for (m in 1:nrow(x)){
      for (n in 1:ncol(x)){
        a <-as.numeric(unlist(strsplit(x[m,n], "|"))[1]) 
        
        b <- as.numeric(unlist(strsplit(x[m,n], "|"))[3])
        gdat[m,n] <- a+b
      }
    }
    rownames(gdat) <- rownames(x)
    colnames(gdat) <- colnames(x)
    return(gdat)
  }
  
  
  #convert to snpMatrix
  gdat <- convertToNumeric(snpMat)
  gdat[1:10, 1:10]
  
  
  ##############################################################################
  #re-label the column names so that they are the locus name
  head(vcf_df)
  snpNames <- vcf_df$ID
  colnames(gdat) <- snpNames
  gdat[1:5, 1:5]
  
  pos_vector <- as.numeric(as.character(vcf_df$POS))
  head(pos_vector)
  
  # save the gdat matrix as a SNPMatrix object, so LDheatmap can work with it
  gdat_snpmat<-as(gdat,"SnpMatrix")
  head(gdat_snpmat)
  gdat_snpmat[1:5, 1:5]
  
  
  ##############################################################################
  # Plot LD using LDheatmap!
  # Open a pdf file
  pdf(paste(fileName, "LD.pdf")) 
  
  #mycols <-viridis_pal(direction = -1)(10) #viridis colors
  mycols <- palette(brewer.pal(n = 9, name = "Blues")) #blue color ramp
  
  LDheatmap(gdat_snpmat,pos_vector,title=fileName,add.map = FALSE,
            LDmeasure="r", color = rev(mycols), flip = TRUE)
  
  dev.off()
}

```

```{r Plot LD Decay and Summary Stats}

#####################################################################################
# specify the relative path to the directory of files
my_path <-"C:/Users/Anita\ Wray/Desktop/AwrayRockfish/populations/refmap/black/LD"
# specify output filenames

output_fileName <- "results_LDdecay_all"
######################################################################################
#To run this code, put all of your vcf files in a single directory

setwd(my_path)
list.files()

# Specify the names of data files used
fileNames <- Sys.glob("*_results.txt") #this is R's version of a wildcard


#read in vcf files in directory using vcfR, and start data processing
i = 1 #initialize counter - for checking loop

df_total = data.frame() #intialize empty dataframe


for (fileName in fileNames) {
  print(i) #counter
  df <- read.delim(fileName)
  df_total <- rbind(df_total, df) # add each individual dataframe to a big dataframe
  i = i +1 #add to counter
}


# Check output 
head(df_total)

## do some magic to assign a numeric value to each chromosome
df_total$CHROM <- as.factor(df_total$CHROM)
chrom <- levels(df_total$CHROM)
chrom


chrom_num <- seq(1,length(chrom), by = 1)
class(chrom_num)

temp_df <- as.data.frame(cbind(chrom, chrom_num))
temp_df


df_total <- left_join(df_total, temp_df, by = c("CHROM" = "chrom"))
head(df_total)

# Save a factor that has information on the numeric number of chromosomes (rockfish have 24)
df_total$chrom_num <- factor(df_total$chrom_num,
                             levels = c("1", "2", "3", "4", "5",
                                        "6", "7", "8", "9", "10", 
                                        "11", "12", "13", "14", "15",
                                        "16", "17", "18", "19", "20",
                                        "21", "22", "23", "24"))

# subset the output a little bit for quick plotting of raw data

df_plot <- df_total %>%
  filter(r2 > 0.005) %>%
  mutate(distance_Mb = distance/1000000)


plot1<- ggplot(df_plot) +
  geom_point(aes(x = distance_Mb, y = r2, color = chrom_num), alpha = 0.2)+
  scale_color_discrete(name = "Chromosome")+
  labs(x="Distance (Mb)",y=expression(LD~(r^{2})))+
  facet_wrap(~chrom_num)+
  theme_bw()+
  theme(legend.position = "none")


plot1
ggsave('LD_per_chromosome.png', plot = plot1)



## summary stats

sum_tbl <- df_total %>%
  filter(r2 > 0.1) %>%
  group_by(CHROM) %>%
  tally() %>%
  mutate(total = sum(n)) %>%
  mutate(freq = n/total)

write.table(sum_tbl, 'summary_table.txt')

```


